### ISSUE TITLE
Pb using C++ in libjit (__dso_handle)

### ISSUE THREAD

**Issue by tlepley-cadence** [pytorch/glow#3839]
We've got some issues when using C++ more extensively in libjit. In particular, it seems that C++ requires a `__dso_handle` symbol (per loaded dynamic lib ?) for static C++ objects (that may simply be pulled when including a STL header).

Apparently, `__dso_handle` is needed to destruct static objects in the right order and it is generally provided by the compiler runtime. 

@opti-mix Any idea how to make this work with the Glow JIT ?

FYI @aturetsk @vuzelac-cadence

**Comment by opti-mix** [comment]
@tlepley-cadence 

> We've got some issues when using C++ more extensively in libjit. In particular, it seems that C++ requires a __dso_handle symbol (per loaded dynamic lib ?) for static C++ objects (that may simply be pulled when including a STL header).
Apparently, __dso_handle is needed to destruct static objects in the right order and it is generally provided by the compiler runtime.

Hmm. I haven't used it in such advanced scenarios yet ;-)

But if you look at the `ExecutionUtils.h`, there are some interesting classes and comments about this topic:
https://llvm.org/doxygen/ExecutionUtils_8h_source.html

Please let me know if it is of any help.

**Comment by tlepley-cadence** [comment]
@opti-mix Thanks for pointing me to this information. It seems that the construction and destruction of static C++ object (that are usually managed by the compiler runtime) need to be handled explicitly by the JIT engine. Utility classes of `ExecutionUtils.h` serve this purpose.

I did some experiment by using `ExecutionUtils.h` classes, being also inspired by the code I've seen in `lli` (https://github.com/microsoft/DirectXShaderCompiler/tree/master/tools/lli) and I managed to get something that seems to work in the simple cases I've tried. This is very experimental code and I'd like to get your point of view since since I guess you are much more familiar with the LLVM JIT. I'm going to create a WIP PR for this purpose.
